### PR TITLE
Show culprit in `setState in render()` fix #12687

### LINKED ISSUE #12687: Show culprit in `Cannot update during an existing state transition...` ?
**Do you want to request a *feature* or report a *bug*?**
*Feature.*

**What is the current behavior?**
We are probably all familiar with the following message.

```
bundle.js:1

Warning: Cannot update during an existing state transition (such as within `render`
or another component's constructor). Render methods should be a pure function
of props and state; constructor side-effects are an anti-pattern, but can be moved
to `componentWillMount`.
```

For those who never came across this message, it can be created simply by mounting the following component.

```js
class UpdateDuringRender extends Component {
  constructor(props) {
    super(props);
    this.state = {
      text: 'initial',
    };
  }

  updateState(text) {
    this.setState({ text });
  }

  render() {
    this.updateState('updated!');

    return (
      <div>
        {this.state.text}
      </div>
    );
  }
}
```

**What is the expected behavior?**
Can we be more specific which component triggers this warning (I think it could be even treated as an optional error)? When this suddenly appears in the app, sometimes it's tough to find the culprit. Some stack trace or at least the offending component would be helpful in the warning message. Not sure how it's solved internally, but yeah, would be a nice hint to help remove anti-patterns.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
`"react": "^16.3.1"`

### PR DESCRIPTION
**Pull request by asiniy**
Fixes #12687

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberScheduler.js [hunk 1]
```diff
@@ -180,30 +178,33 @@ if (__DEV__) {
     didWarnStateUpdateForUnmountedComponent[componentName] = true;
   };
 
-  warnAboutInvalidUpdates = function(instance: React$Component<any>) {
+  warnAboutInvalidUpdates = function(fiber: Fiber) {
+    const componentName = getComponentName(fiber) || 'ReactClass';
     switch (ReactDebugCurrentFiber.phase) {
       case 'getChildContext':
-        if (didWarnSetStateChildContext) {
+        if (didWarnSetStateChildContext[componentName]) {
```
Comment (asiniy) [inline packages/react-reconciler/src/ReactFiberScheduler.js]: Did it for child context as well

File: packages/react-dom/src/__tests__/ReactCompositeComponent-test.js [hunk 2]
```diff
@@ -428,7 +428,8 @@ describe('ReactCompositeComponent', () => {
       'Cannot update during an existing state transition (such as within ' +
         "`render` or another component's constructor). Render methods should " +
         'be a pure function of props and state; constructor side-effects are ' +
-        'an anti-pattern, but can be moved to `componentWillMount`.',
+        'an anti-pattern, but can be moved to `getDerivedStateFromProps`.\n' +
```
Comment (asiniy) [inline packages/react-dom/src/__tests__/ReactCompositeComponent-test.js]: get rid of `componentWillMount`

File: packages/react-reconciler/src/ReactFiberScheduler.js [hunk 3]
```diff
@@ -180,30 +178,33 @@ if (__DEV__) {
     didWarnStateUpdateForUnmountedComponent[componentName] = true;
   };
 
-  warnAboutInvalidUpdates = function(instance: React$Component<any>) {
+  warnAboutInvalidUpdates = function(fiber: Fiber) {
+    const componentName = getComponentName(fiber) || 'ReactClass';
```
Comment (gaearon) [inline packages/react-reconciler/src/ReactFiberScheduler.js]: Let’s change this to “Unknown”? I know it’s not consistent but I like that better

File: packages/react-dom/src/__tests__/ReactCompositeComponent-test.js [hunk 4]
```diff
@@ -475,7 +476,8 @@ describe('ReactCompositeComponent', () => {
     expect(() => {
       instance = ReactDOM.render(<Component />, container);
     }).toWarnDev(
-      'Warning: setState(...): Cannot call setState() inside getChildContext()',
+      'Warning: setState(...): Cannot call setState() inside getChildContext().\n' +
+        '    in Component (at **)',
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactCompositeComponent-test.js]: Let’s rename it in the test so it’s more obvious we have it in the stack. Like MyComponent

File: packages/react-dom/src/__tests__/ReactCompositeComponent-test.js (hunk 2 above)
Comment (gaearon) [reply packages/react-dom/src/__tests__/ReactCompositeComponent-test.js]: Side effects *cannot* be moved to gDSFP, it has to be pure. Let’s revert this part for now? We can think about this message separately

**Comment by gaearon** [comment]
Ping :-)

**Pull request closed without merging**